rfcs/0051-limit-zkapp-cmds-per-block.md

```diff
@@ -0,0 +1,156 @@
+## Summary
+
+During the ITN stress testing it was noticed that daemon's memory
+consumption tends to increase dramatically after a block containing a
+large number of zkApp commands. Before appropriate optimizations can
+be developed, we need a temporary solution to prevent nodes crashing
+due to insufficient memory. The idea is to limit the number of zkApp
+commands that can be included in any single block.
+
+## Motivation
+
+By limiting the number of zkApp commands going into blocks we avoid
+the aforementioned issue until a proper solution can be devised and
+implemented. The root cause of the issue is that proofs contained
+within these commands are stored in the scan state and tend to occupy
+a lot of space. Fixing these storage issues won't affect the
+protocol, so ideally we want a workaround that doesn't affect the
+protocol either, so that at the convenient time we can turn it off
+without making a fork.
+
+## Detailed design
+
+Since the solution should not affect the protocol, it should be
+implemented at the mempool/block producer boundary. In the mempool
+there is `transactions` function, which returns a sequence of
+transactions from the mempool in the order of decreasing transaction
+fees. The block producer module just takes a number of transactions
+from that sequence and includes them. It is relatively easy to limit
+the number of zkApp commands returned by this functions.  ZkApp
+commands above a certain limit would simply be discarded (from the
+sequence, but not from the mempool).
+
+The exact number of zkApps allowed in each block should be set
+dynamically, so that we can adjust it without redeploying nodes.
+The easiest way to control it is through a command line argument
+to the node. Thus changing the setting is as easy as restarting
+the node.
+
+The setting can be stored in the mempool configuration and
+initialized when the mempool is being created at startup.
+It will likely involve some plumbing to transport the setting
+from the command line arguments to the function which initializes
+the mempool, but other than that the solution is quite
+straightforward to implement.
+
+## Drawbacks
```

Comment:
Another drawback is that any payment/stake delegation transactions that's dependent (uses the same fee payer) on a zkApp transaction won't make it until the zkApp transaction is included (which could be delayed because of the limit)